### PR TITLE
Refactor Interactive Brokers Client

### LINKED ISSUE #1397: Refactor the Interactive Brokers Client
# Feature Request

The `InteractiveBrokersClient` serves as the central component of the Interactive Brokers adapter, overseeing a range of critical functions. These include establishing and maintaining connections, handling API errors, executing trades, and gathering various types of data such as market, contract/instrument, and account details. Given its important role and many functions, the Client needs to be particularly well-designed and documented, so that it can be easily understood by new users and maintainers.

As I’ve begun to design strategies in Nautilus Trader, I’ve found that I’ve spent considerable time trying to understand and troubleshoot the behavior of the Client. This is in large part due to the design of IB’s `ibapi` package and confusing elements of the TWS API. That said, there are opportunities to continue to build on the great work of (primarily) @rsmb7z  and @limx0 to make the `InteractiveBrokersClient` even better.

The following are some observations / areas I think can be improved:
- The [class](https://github.com/nautechsystems/nautilus_trader/blob/master/nautilus_trader/adapters/interactive_brokers/client/client.py) is large at 1600+ lines and handles too much functionality for just one class (orders, market data, account management, error handling, connection handling, etc.)
- Some variable and function names can be more descriptive, so their purpose is better understood
- Some logging messages can be added and/or be more descriptive
- There is some repeatable code / patterns that can be further abstracted
- Some code can be refactored (e.g. error handling)
- Some type hints are missing
- Some docstrings are missing
- Many tests are currently skipped or not working

I propose refactoring the `InteractiveBrokersClient` to address these issues / opportunities. The most significant change I propose is splitting the Client class into 7 classes (1 Client + 6 subcomponents), following the principle of composition. Each class should handle a subset of functionality:
- `InteractiveBrokersClient`: This class interfaces with the Interactive Brokers TWS or Gateway and integrates various managers and handlers to provide functionality for connection management, account management, market / instrument data, and order processing.
- `InteractiveBrokersConnectionManager`: This class is dedicated to managing the connection with TWS/Gateway.
- `InteractiveBrokersErrorHandler`: It focuses on addressing all encountered errors and warnings.
- `InteractiveBrokersAccountManager`: Responsible for handling requests related to account information and positions.
- `InteractiveBrokersContractManager`: Handles retrieving contracts (instruments) data
- `InteractiveBrokersMarketDataManager`: Handles market data requests, subscriptions and data processing
- `InteractiveBrokersOrderManager`: Oversees all aspects of order placement and management.

Refactoring the Client will hopefully enable the NT IB adapter maintainers to more easily address issues and add new features. And it should also enable new NT users to more easily investigate issues on their own.

### PR DESCRIPTION
**Pull request by benjaminsingleton**
# Pull Request

Resolves #1397. This is a major refactoring of the InteractiveBrokersClient. This PR is not designed to change the functionality of the DataClient, ExecutionClient, or InstrumentProvider -- those are out of scope.

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## How has this change been tested?

This is a significant change to the core component of the IB adapter, so it warrants thorough review and testing. This PR includes a number of new tests, but it is still far from comprehensive given the broad functionality of Interactive Brokers. Users running strategies using the IB adapter should conduct their own careful testing.

### CODE REVIEW

File: nautilus_trader/adapters/interactive_brokers/client/client.py [hunk 1]
```diff
@@ -268,461 +394,242 @@ def _on_task_completed(
...
         self.is_ready.set()
 
-    def _resume(self):
-        self.is_ready.set()
-        self._connection_attempt_counter = 0
+    def _stop(self) -> None:
+        """
+        Stop the client and cancel running tasks.
 
-    def _degrade(self):
```
Comment (cjdsellers) [inline nautilus_trader/adapters/interactive_brokers/client/client.py:293]: For instance, calling `.degrade()` on the client would eventually call this `_degrade()` method (we should double check the component base class and intended usage).

File: nautilus_trader/adapters/interactive_brokers/client/contract.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+from ibapi.common import SetOfFloat
+from ibapi.common import SetOfString
+from ibapi.contract import ContractDetails
+from ibapi.utils import current_fn_name
+
+from nautilus_trader.adapters.interactive_brokers.common import IBContract
+from nautilus_trader.adapters.interactive_brokers.common import IBContractDetails
+
+
+if TYPE_CHECKING:
```
Comment (cjdsellers) [inline nautilus_trader/adapters/interactive_brokers/client/contract.py:28]: Interesting why we need conditional `TYPE_CHECKING` here?

**Comment by rsmb7z** [comment]
Hi,
I appreciate the focus on adhering to coding standards and recognize the importance of improving the code. However, it's crucial to remember that our ultimate goal is the functional utility of the final product. From my experience with IBAPI, I'd like to emphasize a few key points:

**Gateway Automation/Docker Use:** The use of Gateway Automation or Docker should not be assumed as mandatory. For testing purposes, some may prefer using TWS (Trader Workstation) over Gateway. Similarly, in production environments, opting for out-of-band management instead of the built-in Docker solution can be a viable choice.

**Real-World Testing Scenarios:** Quick functional tests might work well in controlled environments, but real-life scenarios can present unexpected challenges. For example, downloading extensive data for over 500 tickers could reveal issues not apparent in smaller-scale tests.

**Existing Tests and New Changes:** While the current tests, especially those focused on parsing, are limited, they are effective and should continue to pass with any new changes. This ensures that we maintain consistency in our outcomes. For areas lacking tests, we may need to rely on user feedback to identify and address any issues that arise post-implementation.

In conclusion, while we strive for code quality and adherence to standards, let's also keep in mind the diverse ways our software is used in real-world scenarios, and ensure that it remains robust and adaptable to various user needs.

File: .gitignore [hunk 3]
```diff
@@ -60,3 +60,4 @@ bench_data/
 
 !tests/integration_tests/adapters/betfair/responses/*.log
 **/test_db.sqlite
+.cursorignore
```
[the remark this answers is not in the document]
Comment (benjaminsingleton) [reply .gitignore:63]: It's associated with the https://cursor.sh/ browser -- an alternative to GitHub Copilot that I've been playing with. I forgot that I had added it to the .gitignore. Let me know if you prefer that I remove it.

File: nautilus_trader/adapters/interactive_brokers/client/client.py [hunk 4]
```diff
@@ -115,94 +93,51 @@ def __init__(
...
+        self.client_id = client_id
+
+        self._eclient: EClient = EClient(wrapper=self)
+        self._internal_msg_queue: asyncio.Queue = asyncio.Queue()
+        self._connection_attempt_counter = 0
 
-        self._client: EClient = EClient(wrapper=self)
-        self._incoming_msg_queue: asyncio.Queue = asyncio.Queue()
+        # Managers and handlers to handle different aspects of the client
+        self.connection_manager = InteractiveBrokersConnectionManager(self)
```
[the remark this answers is not in the document]
Comment (benjaminsingleton) [reply nautilus_trader/adapters/interactive_brokers/client/client.py:110]: In the case of `connection_manager`, you are correct that this is not referenced external to the class. It should have a leading underscore -- fixing now.

All the other "managers" are referenced in other classes (e.g. methods in `order_manager` are called from the `ExecutionClient`, methods in `contract_manager` are called from the `InstrumentProvider`). A typical linter like pylint will throw errors like: `Access to a protected member <method name> of a client class (protected-access)` if I include the leading underscore.

If the NT convention is to use leading underscores when methods/members are not expected to be used by end users (even if they are called external to the class), I'm very happy to make the required changes.

File: nautilus_trader/adapters/interactive_brokers/client/client.py [hunk 5]
```diff
@@ -233,7 +167,7 @@ def create_task(
         """
         log_msg = log_msg or coro.__name__
         self._log.debug(f"Creating task {log_msg}.")
-        task = self._loop.create_task(
+        task = self.loop.create_task(
```
[the remark this answers is not in the document]
Comment (benjaminsingleton) [reply nautilus_trader/adapters/interactive_brokers/client/client.py:170]: `loop` is referenced external to the class as it appears a few times in the `ConnectionManager` class. For example: https://github.com/benjaminsingleton/nautilus_trader/blob/refactor-client/nautilus_trader/adapters/interactive_brokers/client/connection.py#L131

More than happy to add back the leading underscore. Let me know!

File: nautilus_trader/adapters/interactive_brokers/client/client.py [hunk 6]
```diff
@@ -268,461 +394,242 @@ def _on_task_completed(
...
+
         """
         new_id = self._request_id_seq
         self._request_id_seq += 1
         return new_id
 
-    def _reset(self):
-        self._stop()
-        self._client.reset()
+    def start(self) -> None:
```
[the remark this answers is not in the document]
Comment (benjaminsingleton) [reply nautilus_trader/adapters/interactive_brokers/client/client.py:410]: Ah, I didn't realize that. If that's the case, then I'll add back the leading underscores despite those methods being referenced outside the class. I don't want to introduce any unexpected behavior even if it seems like it wasn't causing issues during testing.

File: nautilus_trader/adapters/interactive_brokers/client/client.py (hunk 1 above)
Comment (benjaminsingleton) [reply nautilus_trader/adapters/interactive_brokers/client/client.py:293]: Noted - adding back the leading underscore.

File: nautilus_trader/adapters/interactive_brokers/client/contract.py (hunk 2 above)
Comment (benjaminsingleton) [reply nautilus_trader/adapters/interactive_brokers/client/contract.py:28]: It's to avoid a circular reference -- `contract.py` needs to import the `InteractiveBrokersClient` from `client.py` but `client.py` needs to import the `InteractiveBrokersContractManager` from `contract.py`.

File: .gitignore (hunk 3 above)
[the remark this answers is not in the document]
Comment (cjdsellers) [reply .gitignore:63]: Thanks cool, it can stay in the gitignore, but I'd probably move it up from the patterns /extras section to the specific file ignores.

File: nautilus_trader/adapters/interactive_brokers/client/client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cjdsellers) [reply nautilus_trader/adapters/interactive_brokers/client/client.py:110]: That's fine, your approach is correct there - lets avoid the access warning and if a class is intended to be called externally then the attribute name doesn't need the leading underscore.

File: nautilus_trader/adapters/interactive_brokers/client/client.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cjdsellers) [reply nautilus_trader/adapters/interactive_brokers/client/client.py:170]: That's fine if its referenced externally, I can't really visualize the design here yet though. Maybe doing a quick diagram with something like https://asciiflow.com/ could be helpful (allows easy editing as its all just copy paste), we could even add it into the docs - just an idea.

File: nautilus_trader/adapters/interactive_brokers/client/client.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cjdsellers) [reply nautilus_trader/adapters/interactive_brokers/client/client.py:410]: So the methods shouldn't be referenced outside the class though, if they are the overrides for those commands being called on the class. Maybe have a quick hunt around for other matching references to compare with other components.

So I might be missing something here myself, but potentially they're being used outside the intended spec?

File: nautilus_trader/adapters/interactive_brokers/client/contract.py (hunk 2 above)
Comment (cjdsellers) [reply nautilus_trader/adapters/interactive_brokers/client/contract.py:28]: Could this be solved by changing the structure of the subpackage and shuffling some modules instead?

**Comment by cjdsellers** [comment]
Further to comments from @rsmb7z

I know some users have this running in production, and so stability is a concern at that point. 

I see @benjaminsingleton intends on adding some more tests which will help with this, and provide some protection against regressions.

It's also great when an adapter is seeing heavy usage from users - because it typically means bugs are uncovered and wrinkles get ironed out quickly.

**Comment by benjaminsingleton** [comment]
Here is a diagram to better visualize the class dependencies. 

The `ExecutionClient`, `DataClient`, `InstrumentProvider`, and `HistoricDataClient` are all directly linked to the `InteractiveBrokersClient` -- they are *not* directly connected to the new classes introduced by this PR. The new classes are exclusively used by the InteractiveBrokersClient.

```
                                                                                              ┌──────────────────────────────────────┐
                                                                                       ┌──────┤ InteractiveBrokersConnectionManager  │
                                                                                       │      ├──────────────────────────────────────┤
                                                                                       │      │                                      │
                                                                                       │      └──────────────────────────────────────┘
       ┌──────────────────────────────────────┐                                        │
┌──────┤ InteractiveBrokersExecutionClient    ├──────┐                                 │
│      ├──────────────────────────────────────┤      │                                 │
│      │                                      │      │                                 │      ┌──────────────────────────────────────┐
│      │                                      │      │                                 │ ┌────┤ InteractiveBrokersAccountManager     ├────┐
│      │                                      │      │                                 │ │    ├──────────────────────────────────────┤    │
│      └──────────────────────────────────────┘      │                                 │ │    │                                      │    │
│                                                    │                                 │ │    └──────────────────────────────────────┘    │
│                                                    │                                 │ │                                                │
│                                                    │                                 │ │                                                │
│      ┌──────────────────────────────────────┐      │  ┌──────────────────────────┐   │ │                                                │
│ ┌────┤ InteractiveBrokersDataClient         ├──┐   │  │ InteractiveBrokersClient │   │ │    ┌──────────────────────────────────────┐    │
│ │    ├──────────────────────────────────────┤  │   │  ├──────────────────────────┤   │ │ ┌──┤ InteractiveBrokersOrderManager       ├────┤
│ │    │                                      │  │   │  │                          │   │ │ │  ├──────────────────────────────────────┤    │
│ │    │                                      │  │   │  │ self._connection_manager ├───┘ │ │  │                                      │    │
│ │    │                                      │  │   │  │                          │     │ │  └──────────────────────────────────────┘    │
│ │    └──────────────────────────────────────┘  │   ├──┤ self.account_manager     ├─────┘ │                                              │
│ │                                              │   │  │                          │       │                                              │
│ │                                              │   └──┤ self.order_manager       ├───────┘                                              │
│ │    ┌──────────────────────────────────────┐  │      │                          │          ┌──────────────────────────────────────┐    │
│ │ ┌──┤ InteractiveBrokersHistoricDataClient ├──┴──────┤ self.market_data_manager ├──────────┤ InteractiveBrokersMarketDataManager  │    │
│ │ │  ├──────────────────────────────────────┤         │                          │          ├──────────────────────────────────────┤    │
│ │ │  │                                      │  ┌──────┤ self.contract_manager    ├───────┐  │                                      │    │
│ │ │  │                                      │  │      │                          │       │  └──────────────────────────────────────┘    │
│ │ │  │                                      │  │      │ self._error_handler      ├────┐  │                                              │
│ │ │  └──────────────────────────────────────┘  │      │                          │    │  │                                              │
│ │ │                                            │      └──────────────────────────┘    │  │                                              │
│ │ │                                            │                                      │  │  ┌──────────────────────────────────────┐    │
│ │ │  ┌──────────────────────────────────────┐  │                                      │  └──┤ InteractiveBrokersContractManager    │    │
└─┴─┴──┤ InteractiveBrokersInstrumentProvider ├──┘                                      │     ├──────────────────────────────────────┤    │
       ├──────────────────────────────────────┤                                         │     │                                      │    │
       │                                      │                                         │     └──────────────────────────────────────┘    │
       │                                      │                                         │                                                 │
       │                                      │                                         │                                                 │
       └──────────────────────────────────────┘                                         │                                                 │
                                                                                        │     ┌──────────────────────────────────────┐    │
                                                                                        └─────┤ InteractiveBrokersErrorHandler       ├────┘
                                                                                              ├──────────────────────────────────────┤
                                                                                              │                                      │
                                                                                              └──────────────────────────────────────┘
```

File: nautilus_trader/adapters/interactive_brokers/client/client.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (benjaminsingleton) [reply nautilus_trader/adapters/interactive_brokers/client/client.py]: I moved a few methods in the `InteractiveBrokersConnectionManager` back to the `InteractiveBrokersClient`. As a result, I've added back the leading underscores for `_start`, `_stop`, etc. 

The `_reset` method in the Client is called from one method in the ConnectionManager, but I've kept the leading underscore so as not to override the `reset` method in the Component base class.

File: nautilus_trader/adapters/interactive_brokers/client/contract.py (hunk 2 above)
Comment (benjaminsingleton) [reply nautilus_trader/adapters/interactive_brokers/client/contract.py:28]: I'm not sure that this would be easy to do due to the way IB's `EClient` and `EWrapper` classes work, but if you or @rsmb7z have thoughts, I'm open ears and happy to incorporate the feedback. Here's my thinking / understanding:

The reason this is tricky is because the `InteractiveBrokersClient` is the class that holds the connection to the TWS API (via the `EWrapper` (to receive messages from the TWS API via websocket) and `EClient` (to send messages to the TWS API via websocket). All the new subcomponents, which are almost entirely comprised of methods simply pulled out of the `Client` to reduce the class size, need to send messages to the TWS API. As a result, the `Client` is passed to the subcomponents as a parameter, but as noted in an earlier comment, the `Client` also instantiates the new subcomponents, so it can delegate to them.

This is a circular dependency, which means that there's a high level of coupling between classes, but in this case, I think it's hard to avoid.

Here is a simplified / contrived example that illustrates what's going on:

```
class InteractiveBrokersClient(Component, EWrapper):
    def __init__(self, ...):
        self._eclient = EClient(wrapper=self)
        self.market_data_manager = InteractiveBrokersMarketDataManager(self)

    # EWrapper override to receive historical bars
    def historicalData(self):
        # delegate to the subcomponent
        self.market_data_manager.receive_bars()

class InteractiveBrokersMarketDataManager:
    def __init__(self, client: InteractiveBrokersClient):
        self._client = client

    def request_bars(self, ...):
        return self._client._eclient.reqHistoricalData(...)

    def receive_bars(self):
        # do stuff then ask for more bars
        self.request_bars(...)
```

I'm going to continue to think about whether there is a way to decouple the dependency.

**Comment by benjaminsingleton** [comment]
@rsmb7z Thanks for sharing your feedback. Here are some thoughts / reactions:

**Gateway Automation/Docker Use**: Fully agree that we should not be designing around an expectation of a dockerized IB gateway. I've had some success with that approach but also challenges working with it.

**Real-World Testing Scenarios**: Agreed. It might be worthwhile to come up with a variety of very basic strategies across different instrument types and market data types to test the adapter in real-life conditions. For example, I run a simple strategy that simply buys then sells every 5 seconds to test the adapter during development. But I haven't yet run the strategy with different instruments/market data, etc.

**Existing Tests and New Changes**: Also in agreement.

File: nautilus_trader/adapters/interactive_brokers/client/contract.py (hunk 2 above)
Comment (benjaminsingleton) [reply nautilus_trader/adapters/interactive_brokers/client/contract.py]: I agree with your suggestion -- mixins are better in this case. I refactored the new subcomponent classes into mixins, and I think this eliminates the circular dependencies and is easier to understand.

**Comment by benjaminsingleton** [comment]
Update: I refactored the new client subcomponent classes into mixins, which reduces complexity and circular dependencies. I need to fix the tests accordingly and do comprehensive live testing. Still a work in progress!

**Comment by cjdsellers** [comment]
Hey @benjaminsingleton 

Just a heads up there was a slight interface change from this diff ab5170fc5311e4ff0c046a061d1ee88bc67d782a for instrument requests you'll need to incorporate. These requests now include a `start` and `end` which was needed for the Databento integration (and has better symmetry with the other requests now too).

**Comment by benjaminsingleton** [comment]
Update -- Continuing to refactor and wrote a significant number of tests. Still some more work to do but getting close to ready.

File: tests/integration_tests/adapters/interactive_brokers/conftest.py [hunk 7]
```diff
@@ -73,9 +73,9 @@ def exec_client_config():
 
 
 @pytest.fixture()
-def ib_client(data_client_config, loop, msgbus, cache, clock, logger):
+def ib_client(data_client_config, event_loop, msgbus, cache, clock, logger):
```
Comment (cjdsellers) [inline tests/integration_tests/adapters/interactive_brokers/conftest.py:76]: Small nit but the event loop param is very typically `loop`.

Comment (benjaminsingleton) [reply tests/integration_tests/adapters/interactive_brokers/conftest.py:76]: Will revert!

**Comment by benjaminsingleton** [comment]
Tests are now written, so this is ready for another review. As I noted in the PR description, the tests are certainly not comprehensive, but I think they provide a solid foundation. We can revisit tests for the DataClient and ExecutionClient in a subsequent PR.

I'd love to ask some folks to test run their strategies to ensure the Client is behaving as expected. What's the best way to do that @cjdsellers? Make a request in Discord?

**Comment by cjdsellers** [comment]
Hey @benjaminsingleton 

I must have missed this message - yes I think thats a good idea.
@rsmb7z any thoughts or feeback on this one now?

**Comment by cjdsellers** [comment]
I've had a brief skim and its looking great to me, nothing glaringly wrong which I can see - nice job!

I'd be happy to merge this post next release (very likely tomorrow), and give people a chance to try it out on `develop` and `nightly` branches.

**Comment by cjdsellers** [comment]
Lets go for it - thanks so much again :pray: @benjaminsingleton

**Comment by benjaminsingleton** [comment]
Great! Once this is merged into develop/nightly, I'll make a post in the Discord to ask folks to send feedback, and I'll patch up whatever is needed.
